File: leetcode/python/797_all_paths_from_source_to_target.py
class Solution:
    def allPathsSourceTarget(self, graph: list[list[int]]) -> list[list[int]]:
        res: list[list[int]] = []
        self.dfs(graph, 0, [], res)
        return res
    
    def dfs(self, graph: list[list[int]], node: int, path: list[int], res: list[list[int]]) -> None:
        path.append(node)
        if node == len(graph) - 1:
            res.append(path.copy())
            path.pop()
            return
        
        for neighbor in graph[node]:
            self.dfs(graph, neighbor, path, res)
        path.pop()

if __name__ == '__main__':
    s = Solution()
    print(s.allPathsSourceTarget([[1,2],[3],[3],[]]))
    print(s.allPathsSourceTarget([[4,3,1],[3,2,4],[3],[4],[]]))

File: leetcode/c++/797_all_paths_from_source_to_target.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        vector<int> path = {0};
        dfs(graph, 0, path);
        return res;
    }

    void dfs(vector<vector<int>>& graph, int node, vector<int>& path) {
        if (node == graph.size() - 1) {
            res.push_back(path);
            return;
        }
        for (int i = 0; i < graph[node].size(); i++) {
            path.push_back(graph[node][i]);
            dfs(graph, graph[node][i], path);
            path.pop_back();
        }
    }
private:
    vector<vector<int>> res;
};